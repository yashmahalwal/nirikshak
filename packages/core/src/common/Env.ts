/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
