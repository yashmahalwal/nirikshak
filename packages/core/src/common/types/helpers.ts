import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
