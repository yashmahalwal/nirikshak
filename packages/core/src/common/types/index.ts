export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
