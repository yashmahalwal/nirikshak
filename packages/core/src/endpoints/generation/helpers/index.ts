export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
