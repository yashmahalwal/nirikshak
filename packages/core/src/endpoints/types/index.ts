export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
