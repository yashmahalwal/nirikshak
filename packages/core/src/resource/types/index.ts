import {
    isBaseType,
    isWithModifiersBaseType,
    isWithModifiersResource,
    isOneOfEntries,
    BaseType,
    WithModifiers,
    OneOfEntries,
} from "./helper";

// type guard: ResourceBase
export function isResourceBase(input: any): input is ResourceBase {
    if (!input || typeof input != "object" || Array.isArray(input))
        return false;
    return Object.keys(input).every(
        (key) =>
            isBaseType(input[key]) ||
            isWithModifiersBaseType(input[key]) ||
            isWithModifiersResource(input[key]) ||
            isOneOfEntries(input[key]) ||
            isResourceBase(input[key])
    );
}

// A basic key value pair
export interface ResourceBase {
    [key: string]:
        | BaseType
        | WithModifiers<BaseType>
        | WithModifiers<ResourceBase>
        | OneOfEntries
        // Check for resource base at last as it matches any key
        | ResourceBase;
}

// A resource schema is a resource base along with an identifier field
export interface Resource extends ResourceBase {
    id: Exclude<BaseType, Array<any> | null | boolean>;
}

// type guard: isResource
export function isResource(input: any): input is Resource {
    if (!input || typeof input !== "object") return false;

    return (
        !Array.isArray(input["id"]) &&
        isBaseType(input.id) &&
        isResourceBase(input)
    );
}
