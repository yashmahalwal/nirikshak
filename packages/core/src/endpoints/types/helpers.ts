import { ResourceString, isResourceString } from "./resourceString";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { Primitives } from "../../common/types/helpers";

export type MethodType = "DELETE" | "GET" | "PUT" | "POST" | "PATCH";
export function isMethodType(key: any): key is MethodType {
    if (typeof key !== "string") return false;

    switch (key) {
        case "DELETE":
        case "GET":
        case "PUT":
        case "POST":
        case "PATCH":
            return true;
        default:
            return false;
    }
}

export type Cases = "NEGATIVE" | "POSITIVE" | "DESTRUCTIVE";

export function isCaseType(key: any): key is Cases {
    if (typeof key !== "string") return false;
    switch (key) {
        case "NEGATIVE":
        case "POSITIVE":
        case "DESTRUCTIVE":
            return true;
        default:
            return false;
    }
}

export type BaseType =
    | FakerType
    | CustomFunctionType
    | ResourceString
    | Literal;

export function isBaseType(input: any): input is BaseType {
    return (
        isFakerType(input) ||
        isCustomFunction(input) ||
        isResourceString(input) ||
        isLiteral(input)
    );
}

export type Base = BaseType | Base[];

export function isBase(input: any): input is Base {
    if (Array.isArray(input)) return input.every((value) => isBase(value));
    return isBaseType(input);
}

export type WithModifiers<T extends Base | BodyType> = (T extends Base
    ? { type: Base }
    : { field: BodyType }) & {
    plural?: boolean;
    nullable?: boolean;
    optional?: boolean;
};

// Checking the validity of the modifier fields of the HOT
function checkModifiers(input: WithModifiers<Base | BodyType>): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersBodyType(
    input: any
): input is WithModifiers<BodyType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isBodyType(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBase(input: any): input is WithModifiers<Base> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBase(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<Base | BodyType> {
    return isWithModifiersBase(input) || isWithModifiersBodyType(input);
}

export type OneOfEntries =
    | { fields: (BodyType | WithModifiers<BodyType>)[] }
    | { types: (Base | WithModifiers<Base>)[] }
    | {
          fields: (BodyType | WithModifiers<BodyType>)[];
          types: (Base | WithModifiers<Base>)[];
      };

// type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBase(entry) || isWithModifiersBase(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) => isBodyType(entry) || isWithModifiersBodyType(entry)
              )
            : true);

    return value;
}

export interface BodyType {
    [key: string]:
        | ResourceString
        | Exclude<Base, ResourceString>
        | WithModifiers<Base>
        | WithModifiers<BodyType>
        | OneOfEntries
        | BodyType;
}

export interface BodyInstance {
    [key: string]: Primitives | BodyInstance | BodyInstance[];
}

export function isBodyType(input: any): input is BodyType {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (entry) =>
            isResourceString(entry) ||
            isBase(entry) ||
            isWithModifiers(entry) ||
            isOneOfEntries(entry) ||
            isBodyType(entry)
    );
}

export interface HeaderMap {
    [key: string]: Exclude<BaseType, null>;
}

export function isHeaderMap(input: any): input is HeaderMap {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (value) => value !== null && isBaseType(value)
    );
}
