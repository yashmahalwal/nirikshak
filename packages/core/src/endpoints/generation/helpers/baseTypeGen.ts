import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers } from "../../../common/types/helpers";
import { Literal, isLiteral } from "../../../common/types/literals";
import { isFakerType } from "../../../common/types/fakerTypes";
import { generateFaker } from "../../../common/generation/fakerGen";
import { isCustomFunction } from "../../../common/types/custom";
import { generateCustom } from "../../../common/generation/customGen";
import { isResourceString } from "../../types/resourceString";
import { getFromResource } from "../resourceStringGen";
import { BaseType } from "../../types/helpers";

export async function generateBaseType(
    input: BaseType,
    instance: ResourceInstance,
    helpers: SchemaHelpers
): Promise<Literal> {
    if (isFakerType(input)) return generateFaker(input);
    if (isCustomFunction(input)) return generateCustom(input, helpers);
    if (isResourceString(input)) {
        const value = getFromResource(input, instance);
        if (!isLiteral(value))
            throw new Error(
                `Resource string ${input} resolved to non primitive value`
            );
        return value;
    }
    return input;
}
