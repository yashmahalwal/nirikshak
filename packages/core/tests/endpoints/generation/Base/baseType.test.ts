import { BaseType } from "../../../../src/endpoints/types/helpers";
import {
    Primitives,
    SchemaHelpers,
} from "../../../../src/common/types/helpers";
import faker from "faker";
import { generateBaseType } from "../../../../src/endpoints/generation/helpers/baseTypeGen";
import { RANDOMNESS_ITERATIONS } from "../../../../src/common/Env";

const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidEntries: { input: BaseType; output: () => Promise<Primitives> }[] = [
    {
        input: 1,
        async output(): Promise<1> {
            return 1;
        },
    },
    {
        input: {
            function: "faker:random.number",
            args: [10],
        },
        async output(): Promise<number> {
            return faker.random.number(10);
        },
    },
    {
        input: "custom:word",
        async output(): Promise<string> {
            return Helpers.word() as Promise<string>;
        },
    },
    {
        input: "resource:addresses[1].zipCode",
        async output(): Promise<string> {
            return resource.addresses[1].zipCode;
        },
    },
];

const InvalidEntries: any[] = [
    "custom:nonExistent",
    "custom:invalidHelper",
    "resource:invalidPath",
    "resource:addresses",
];

describe("BaseType type generation", () => {
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
        ValidEntries.forEach((entry, index) =>
            test(`Valid entry: ${index}, iteration: ${i}`, async () => {
                faker.seed(123 + index * i);
                const got = await generateBaseType(
                    entry.input,
                    resource,
                    Helpers
                );
                faker.seed(123 + index * i);
                const expected = await entry.output();
                expect(got).toEqual(expected);
            })
        );

    test.each(InvalidEntries)(`Invalid entry: %#`, (entry) =>
        expect(
            generateBaseType(entry, resource, Helpers)
        ).rejects.toMatchSnapshot()
    );
});
