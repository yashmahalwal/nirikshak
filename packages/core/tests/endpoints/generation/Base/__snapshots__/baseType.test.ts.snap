// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseType type generation Invalid entry: 0 1`] = `[Error: function: custom:nonExistent, args: ]`;

exports[`BaseType type generation Invalid entry: 1 1`] = `[Error: Non literal type value from custom:invalidHelper]`;

exports[`BaseType type generation Invalid entry: 2 1`] = `[Error: Invalid resource instance path: invalidPath]`;

exports[`BaseType type generation Invalid entry: 3 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;
